packages/app/src/runner/events/capture-protocol.ts

```diff
@@ -0,0 +1,33 @@
+export const addCaptureProtocolListeners = (Cypress: Cypress.Cypress) => {
+  Cypress.on('log:added', (log) => {
+    const displayProps = Cypress.runner.getDisplayPropsForLog(log)
+
+    Cypress.backend('protocol:command:log:added', displayProps)
+  })
+
+  Cypress.on('log:changed', (log) => {
+    const displayProps = Cypress.runner.getDisplayPropsForLog(log)
+
+    Cypress.backend('protocol:command:log:changed', displayProps)
+  })
+
+  Cypress.on('viewport:changed', (viewport) => {
+    const timestamp = performance.timeOrigin + performance.now()
+
+    Cypress.backend('protocol:viewport:changed', {
+      viewport: {
+        width: viewport.viewportWidth,
+        height: viewport.viewportHeight,
+      },
+      timestamp,
+    })
+  })
+
+  Cypress.on('test:before:run:async', async (attributes) => {
+    await Cypress.backend('protocol:test:before:run:async', attributes)
```

Comment:
Ensure we setup the active runnable before the test starts.

Reply:
Since it's just an event emitter is that signaled back to the runner before the test starts? In other words, what is waiting for this promise to settle / waiting for this event to finish?

Reply:
The standard `emit` function is not used for `test:before:run:async`, instead [emitThen](https://github.com/cypress-io/cypress/blob/ddf720482bb0f3a3bc4a630b9c4b270b52fb80de/packages/driver/src/cypress.ts#L547) is used. You can trace that back to the runner where it is [called](https://github.com/cypress-io/cypress/blob/ddf720482bb0f3a3bc4a630b9c4b270b52fb80de/packages/driver/src/cypress/runner.ts#L1651). The test is then run in the [finally](https://github.com/cypress-io/cypress/blob/ddf720482bb0f3a3bc4a630b9c4b270b52fb80de/packages/driver/src/cypress/runner.ts#L1675) block after all of those promises resolve.